test/performance/odp_ipsecfwd.c

```diff
@@ -489,12 +501,26 @@ static inline uint32_t forward_packets(odp_packet_t pkts[], int num, odph_table_
 			odp_packet_free(pkt);
 			continue;
 		}
-		if (odp_unlikely(!pktio->send_fn(pktio, hash % pktio->num_tx_qs, pkt))) {
-			odp_packet_free(pkt);
+
+		vec = &vecs[pktio->idx];
+		vec->pkts[vec->num++] = pkt;
+		vec->hash = hash;
+		vec->pktio = pktio;
+	}
+
+	for (uint32_t i = 0U; i < MAX_IFS; ++i) {
+		if (vecs[i].num == 0U)
 			continue;
-		}
 
-		++num_procd;
+		vec = &vecs[i];
+		pktio = vec->pktio;
+		uint32_t ret = pktio->send_fn(pktio, vec->hash % pktio->num_tx_qs, vec->pkts,
```

Comment:
Hash based queue selection is not needed in direct mode (and is not needed in the unoredered scheduled mode that was already supported) since packet order across threads is not maintained anyway. Thread-specific output queues would be ok and fast (and could be used in the MT_UNSAFE mode) with direct pktio output.

Reply:
New patch (06ba5ae) for this in v3.